### PR TITLE
Support stable scheduling for TiDB

### LINKED ISSUE #332: Proposal: Allow tidb pod to be scheduled on the previous node
For `NodePort` type Service, the `spec.externalTrafficPolicy` may set to `Local` to enable client IP propagation which is needed for source IP based permission control. This way, not all the nodes and node port is accessible for the client connection. Users may choose only the node where the tidb-server pod is on to connect. But when performing a rolling upgrade, the tidb-server pod may be scheduled to other nodes, thus the client connection info needs to be updated. This is inconvenient for users.

So here we propose to allow tidb pod to be scheduled on the previous node. Currently, the tidb pod scheduling will bypass our customized scheduler. This can be done by adding a scheduling policy for the tidb pod. We can record the node IP of the tidb pod in `TidbCluster` status, and give higher priority of the previous node if multiple nodes match tidb pod requirements.

### PR DESCRIPTION
**Pull request by cofyc**
### What problem does this PR solve? 

fixes #332

### What is changed and how it works?

See design proposal in https://github.com/pingcap/tidb-operator/pull/466

### Check List 

Tests 

Unit test (done)
E2E test (done)
Manual test (done)

Code changes

Has Go code change

Side effects

Controlled by a feature gate, no side effects if not enabled.

Related changes

Need to update the documentation

### Does this PR introduce a user-facing change?:

 
 ```release-note
Support stable scheduling for pods of TiDB members in tidb-scheduler.
 ```

### CODE REVIEW

File: pkg/scheduler/predicates/stable_scheduling.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		for _, node := range nodes {
+			if node.Name == nodeName {
+				glog.V(4).Infof("previous node %q for pod %q in TiDB cluster %q exists in candicates, filter out other nodes", nodeName, podName, tcName)
+				return []apiv1.Node{node}, nil
+			}
+		}
+		msg := fmt.Sprintf("cannot run on its previous node %q", nodeName)
+		p.recorder.Event(pod, apiv1.EventTypeWarning, UnableToRunOnPreviousNodeReason, msg)
+	} else {
+		glog.V(4).Infof("no previous node exists for pod %q in TiDB cluster %q", podName, tcName)
```
Comment (weekface) [inline pkg/scheduler/predicates/stable_scheduling.go]:
```suggestion
		glog.V(4).Infof("no previous node exists for pod %q in TiDB cluster %s/%q", podName, ns, tcName)
```
log the namespace?

File: pkg/scheduler/predicates/stable_scheduling.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, err
+	}
+
+	nodeName := p.findPreviousNodeInTC(tc, pod)
+
+	if nodeName != "" {
+		glog.V(4).Infof("found previous node %q for pod %q in TiDB cluster %q", nodeName, podName, tcName)
+		for _, node := range nodes {
+			if node.Name == nodeName {
+				glog.V(4).Infof("previous node %q for pod %q in TiDB cluster %q exists in candicates, filter out other nodes", nodeName, podName, tcName)
```
Comment (weekface) [inline pkg/scheduler/predicates/stable_scheduling.go]: use `glog.V(2)`?

Comment (cofyc) [reply pkg/scheduler/predicates/stable_scheduling.go]: Print these logs by default?

Comment (weekface) [reply pkg/scheduler/predicates/stable_scheduling.go]: Yep

**Comment by weekface** [comment]
please add a release note to https://github.com/pingcap/tidb-operator/pull/477#issue-277329118

**Comment by cofyc** [comment]
done

**Comment by weekface** [comment]
@tennix @xiaojingchen PTAL

**Comment by cofyc** [comment]
@weekface @xiaojingchen I've rebased on the latest master, PTAL again.

**Comment by cofyc** [comment]
/run-e2e-tests
Merge master and run e2e again.

**Comment by cofyc** [comment]
/run-e2e-tests
Merge master and run e2e again.

**Comment by cofyc** [comment]
Follow-up issue: https://github.com/pingcap/tidb-operator/issues/498

**Pull request merged**